File: pools/allocating_strategies/strategies/src/ipv6_prefix_strategy.js
import {
    hostsInMask,
    inet_aton,
    inet_ntoa,
    parsePrefix,
    prefixToStr,
    subnetAddresses
} from "./ipv6-utils";

// framework managed constants
var currentResources = []
var resourcePoolProperties = {}
var userInput = {}
// framework managed constants

// STRATEGY_START

/*
IPv6 prefix allocation strategy

- Expects Ipv6 prefix resource type to have 2 properties of type int ["address:string", "mask:int"]
- userInput.desiredSize is a required parameter
- userInput.subnet is an optional parameter specifying whether allocated prefix will be used as a real subnet or just
  as an IP pool. Essentially whether to count subnet address and broadcast into the range or not. If set to true, the resulting
  prefix will have a capacity of desiredSize + 2. Defaults to false
- Logs utilisation stats
- Allocates previously freed prefixes
- All addresses from parent prefix are used, including the first and last one
 */

// compare prefixes based on their broadcast address
function comparePrefix(prefix1, prefix2) {
    let endOfP1 = inet_aton(prefix1.address) + subnetAddresses(prefix1.prefix)
    let endOfP2 = inet_aton(prefix2.address) + subnetAddresses(prefix2.prefix)
    if (endOfP1 - endOfP2 < BigInt(0)) {
        return -1
    } else if (endOfP1 - endOfP2 > BigInt(0)) {
        return 1
    }

    return 0
}

// sum up capacity of an array of addresses
function prefixesCapacity(currentResources) {
    let width = BigInt(0)
    for (let allocatedPrefix of currentResources) {
        width += subnetAddresses(allocatedPrefix.prefix)
    }
    return width
}

// calculate utilized capacity based on previously allocated prefixes + a newly allocated prefix
function utilizedCapacity(allocatedRanges, newlyAllocatedRangeCapacity) {
    return prefixesCapacity(allocatedRanges) + newlyAllocatedRangeCapacity
}

// calculate free capacity based on previously allocated prefixes
function freeCapacity(parentPrefix, utilisedCapacity) {
    return subnetAddresses(parentPrefix.prefix) - utilisedCapacity;
}


function capacity() {
    let totalCapacity = hostsInMask(resourcePoolProperties.address, resourcePoolProperties.prefix);
    let allocatedCapacity = BigInt(0);
    let resource;
    let subnetItself = userInput.subnet ? BigInt(1) : BigInt(0);

    for (resource of currentResources) {
        allocatedCapacity += hostsInMask(resource.Properties.address, resource.Properties.prefix);
    }

    return {
        freeCapacity: String(totalCapacity - allocatedCapacity + subnetItself),
        utilizedCapacity: String(allocatedCapacity)
    };
}

// log utilisation stats
function logStats(newlyAllocatedPrefix, parentRange, allocatedPrefixes = [], level = "log") {
    let newlyAllocatedPrefixCapacity = 0
    if (newlyAllocatedPrefix) {
        newlyAllocatedPrefixCapacity = subnetAddresses(newlyAllocatedPrefix.prefix)
    } else {
        newlyAllocatedPrefixCapacity = 0
    }

    let utilisedCapacity = utilizedCapacity(allocatedPrefixes, BigInt(newlyAllocatedPrefixCapacity))
    let remainingCapacity = freeCapacity(parentRange, utilisedCapacity)
    let utilPercentage
    if (remainingCapacity === BigInt(0)) {
        utilPercentage = 100.0
    } else {
        let utilFloat = Number(utilisedCapacity * BigInt(1000) / subnetAddresses(parentRange.prefix)) / 1000
        utilPercentage = (utilFloat * 100)
    }
    console[level]("Remaining capacity: " + remainingCapacity)
    console[level]("Utilised capacity: " + utilisedCapacity)
    console[level](`Utilisation: ${utilPercentage.toFixed(1)}%`)
}

function prefixesToString(currentResourcesUnwrapped) {
    let prefixesToStr = ""
    for (let allocatedPrefix of currentResourcesUnwrapped) {
        prefixesToStr += prefixToStr(allocatedPrefix)
        prefixesToStr += prefixToRangeStr(allocatedPrefix)
        prefixesToStr += ", "
    }
    return prefixesToStr
}

function prefixToRangeStr(prefix) {
    return `[${prefix.address}-${inet_ntoa(inet_aton(prefix.address) + subnetAddresses(prefix.prefix) - BigInt(1))}]`
}

function calculateDesiredSubnetMask() {
    let desiredSizeBigInt = BigInt(userInput.desiredSize)
    let newSubnetBits = BigInt(1)
    for (let i = 1; i <= 128; i++) {
        newSubnetBits *= BigInt(2)
        if (newSubnetBits >= desiredSizeBigInt) {
            newSubnetBits = i
            break;
        }
    }
    let newSubnetMask = 128 - newSubnetBits
    let newSubnetCapacity = subnetAddresses(newSubnetMask)
    return {newSubnetMask, newSubnetCapacity}
}

// calculate the nearest possible address for a subnet where mask === newSubnetMask
//  that is outside of allocatedSubnet
function findNextFreeSubnetAddress(allocatedSubnet, newSubnetMask) {
    // find the first address after currently iterated allocated subnet
    let nextAvailableAddressNum = inet_aton(allocatedSubnet.address) + subnetAddresses(allocatedSubnet.prefix)
    // remove any bites from the address above after newSubnetMask
    let newSubnetMaskNegative = BigInt(128) - BigInt(newSubnetMask)
    let possibleSubnetNum = (nextAvailableAddressNum >> newSubnetMaskNegative) << newSubnetMaskNegative
    // keep going until we find an address outside of currently iterated allocated subnet
    while (nextAvailableAddressNum > possibleSubnetNum) {
        possibleSubnetNum = ((possibleSubnetNum >> newSubnetMaskNegative) + BigInt(1)) << newSubnetMaskNegative
    }
    return possibleSubnetNum
}

// main
function invoke() {
    let rootPrefixParsed = resourcePoolProperties
    if (rootPrefixParsed == null) {
        console.error("Unable to extract root prefix from pool name: " + rootPrefix)
        return null
    }
    let rootAddressStr = rootPrefixParsed.address
    let rootMask = rootPrefixParsed.prefix
    let rootPrefixStr = prefixToStr(rootPrefixParsed)
    let rootCapacity = subnetAddresses(rootMask)
    let rootAddressNum = inet_aton(rootAddressStr)

    if (!userInput.desiredSize) {
        console.error("Unable to allocate subnet from root prefix: " + rootPrefixStr +
            ". Desired size of a new subnet size not provided as userInput.desiredSize")
        return null
    }

    // Convert desiredSize to bigint
    userInput.desiredSize = BigInt(userInput.desiredSize)

    if (userInput.desiredSize < BigInt(2)) {
        console.error("Unable to allocate subnet from root prefix: " + rootPrefixStr +
            ". Desired size is invalid: " + userInput.desiredSize + ". Use values >= 2")
        return null
    }

    if (userInput.subnet === true) {
        // reserve subnet address and broadcast
        userInput.desiredSize += BigInt(2)
    }

    // Calculate smallest possible subnet mask to fit desiredSize
    let {newSubnetMask, newSubnetCapacity} = calculateDesiredSubnetMask()

    // unwrap and sort currentResources
    let currentResourcesUnwrapped = currentResources.map(cR => cR.Properties)
    currentResourcesUnwrapped.sort(comparePrefix)

    let possibleSubnetNum = rootAddressNum
    // iterate over allocated subnets and see if a desired new subnet can be squeezed in
    for (let allocatedSubnet of currentResourcesUnwrapped) {

        let allocatedSubnetNum = inet_aton(allocatedSubnet.address)
        let chunkCapacity = allocatedSubnetNum - possibleSubnetNum
        if (chunkCapacity >= userInput.desiredSize) {
            // there is chunk with sufficient capacity between possibleSubnetNum and allocatedSubnet.address
            let newlyAllocatedPrefix = {
                "address": inet_ntoa(possibleSubnetNum),
                "prefix": newSubnetMask
            }
            // FIXME How to pass these stats ?
            // logStats(newlyAllocatedPrefix, rootPrefixParsed, currentResourcesUnwrapped)
            return newlyAllocatedPrefix
        }

        // move possible subnet start to a valid address outside of allocatedSubnet's addresses and continue the search
        possibleSubnetNum = findNextFreeSubnetAddress(allocatedSubnet, newSubnetMask)
    }

    // check if there is any space left at the end of parent range
    if (possibleSubnetNum + newSubnetCapacity <= rootAddressNum + rootCapacity) {
        // there sure is some space, use it !
        let newlyAllocatedPrefix = {
            "address": inet_ntoa(possibleSubnetNum),
            "prefix": newSubnetMask
        }
        // FIXME How to pass these stats ?
        // logStats(newlyAllocatedPrefix, rootPrefixParsed, currentResourcesUnwrapped)
        return newlyAllocatedPrefix
    }

    // no suitable range found
    console.error("Unable to allocate Ipv6 prefix from: " + rootPrefixStr +
        ". Insufficient capacity to allocate a new prefix of size: " + userInput.desiredSize)
    console.error("Currently allocated prefixes: " + prefixesToString(currentResourcesUnwrapped))
    logStats(null, rootPrefixParsed, currentResourcesUnwrapped, "error")
    return null
}

// STRATEGY_END

// For testing purposes
function invokeWithParams(currentResourcesArg, resourcePoolArg, userInputArg) {
    currentResources = currentResourcesArg
    resourcePoolProperties = resourcePoolArg
    userInput = userInputArg
    return invoke()
}

// For testing purposes
function invokeWithParamsCapacity(currentResourcesArg, resourcePoolArg, userInputArg) {
    currentResources = currentResourcesArg
    resourcePoolProperties = resourcePoolArg
    userInput = userInputArg
    return capacity()
}

exports.invoke = invoke
exports.capacity = capacity
exports.invokeWithParams = invokeWithParams
exports.invokeWithParamsCapacity = invokeWithParamsCapacity
exports.utilizedCapacity = utilizedCapacity
exports.freeCapacity = freeCapacity
exports.parsePrefix = parsePrefix
exports.inet_ntoa = inet_ntoa
exports.inet_aton = inet_aton
// For testing purposes

File: pools/allocating_strategies/strategies/src/ipv6_prefix_strategy.go
package src

import (
	"github.com/pkg/errors"
	"math/big"
	"sort"
	"strconv"
)

type Ipv6Prefix struct {
	currentResources       []map[string]interface{}
	resourcePoolProperties map[string]interface{}
	userInput              map[string]interface{}
}

type Ipv6PrefixStruct struct {
	address string
	prefix  int
}

func NewIpv6Prefix(currentResources []map[string]interface{},
	resourcePoolProperties map[string]interface{},
	userInput map[string]interface{}) Ipv6Prefix {
	return Ipv6Prefix{currentResources, resourcePoolProperties, userInput}
}

func (ipv6Prefix *Ipv6Prefix) UtilizedCapacity(allocatedRanges []map[string]interface{}, newlyAllocatedRangeCapacity float64) float64 {
	return float64(len(allocatedRanges)) + newlyAllocatedRangeCapacity
}

// FreeCapacity calculate free capacity based on previously allocated prefixes
func (ipv6Prefix *Ipv6Prefix) FreeCapacity(parentPrefix string, utilisedCapacity float64) float64 {
	capacityString := strconv.FormatFloat(utilisedCapacity, 'f', -1, 64)
	capacityInt, _ := new(big.Int).SetString(capacityString, 10)
	parentPrefixInt, _ := strconv.Atoi(parentPrefix)
	addressesCount := ipv6SubnetAddresses(parentPrefixInt)
	addressesCount.Sub(addressesCount, capacityInt)
	return float64(addressesCount.Int64())
}

func (ipv6Prefix *Ipv6Prefix) Capacity() (map[string]interface{}, error) {
	var result = make(map[string]interface{})
	rootAddressStr, ok := ipv6Prefix.resourcePoolProperties["address"]
	if !ok {
		return nil, errors.New("Unable to extract address resource")
	}
	rootMask, ok := ipv6Prefix.resourcePoolProperties["prefix"]
	if !ok {
		return nil, errors.New("Unable to extract prefix resources")
	}
	subnetItself := new(big.Int)
	i, ok := ipv6Prefix.userInput["subnet"]
	if i != nil {
		subnetItself = big.NewInt(1)
	} else {
		subnetItself = big.NewInt(0)
	}

	var allocatedCapacity = big.NewInt(0)
	for _, resource := range ipv6Prefix.currentResources {
		prefix, err := NumberToInt(resource["Properties"].(map[string]interface{})["prefix"].(interface{}))
		if err != nil {
			return nil, err
		}
		allocatedCapacity.Add(allocatedCapacity, ipv6HostsInMask(
			resource["Properties"].(map[string]interface{})["address"].(string),
			prefix.(int)))
	}
	rootMask, err := NumberToInt(rootMask)
	if err != nil {
	    return nil, err
	}
	totalCapacity := ipv6HostsInMask(rootAddressStr.(string), rootMask.(int))
	totalCapacity.Sub(totalCapacity, allocatedCapacity)
	totalCapacity.Add(totalCapacity, subnetItself)

	result["freeCapacity"] = totalCapacity.String()
	result["utilizedCapacity"] = allocatedCapacity.String()
	return result, nil
}

func (ipv6Prefix *Ipv6Prefix) Invoke() (map[string]interface{}, error) {
	if ipv6Prefix.resourcePoolProperties == nil {
		return nil, errors.New("Unable to extract resources")
	}
	rootAddressStr, ok := ipv6Prefix.resourcePoolProperties["address"]
	if !ok {
		return nil, errors.New("Unable to extract address resource")
	}
	rootMask, ok := ipv6Prefix.resourcePoolProperties["prefix"]
	if !ok {
		return nil, errors.New("Unable to extract prefix resources")
	}
	rootMask, err := NumberToInt(rootMask)
	if err != nil {
		return nil, err
	}
	rootPrefixStr := prefixToStr(ipv6Prefix.resourcePoolProperties)
	rootCapacity := ipv6SubnetAddresses(rootMask.(int))
	rootAddressNum, err := Ipv6InetAton(rootAddressStr.(string))
	if err != nil {
		return nil, err
	}

	value, ok := ipv6Prefix.userInput["desiredSize"]
	if !ok {
		return nil, errors.New("Unable to allocate subnet from root prefix: " + rootPrefixStr +
			". Desired size of a new subnet size not provided as userInput.desiredSize")
	}
	desiredSize, err := NumberToBigInt(value)
	if err != nil {
		return nil, err
	}
	if desiredSize.Cmp(big.NewInt(2)) < 0 {
		return nil, errors.New("Unable to allocate subnet from root prefix: " + rootPrefixStr +
			". Desired size is invalid: " + desiredSize.String() + ". Use values >= 2")
	}

	if value, ok := ipv6Prefix.userInput["subnet"]; ok && value == true {
		// reserve subnet address and broadcast
		desiredSize.Add(desiredSize, big.NewInt(2))
	}

	// Calculate smallest possible subnet mask to fit desiredSize
	newSubnetMask, newSubnetCapacity := calculateDesiredSubnetMask(desiredSize)

	var currentResourcesStruct []Ipv6PrefixStruct
	for _, resource := range ipv6Prefix.currentResources {
		address, prefix, err := getAddressAndPrefixFromCurrentResource(resource)
		if err != nil {
			return nil, err
		}
		currentResourcesStruct = append(currentResourcesStruct, Ipv6PrefixStruct{address: address, prefix: prefix})
	}

	sort.Slice(currentResourcesStruct, func(i, j int) bool {
		address1, _ := Ipv6InetAton(currentResourcesStruct[i].address)
		address2, _ := Ipv6InetAton(currentResourcesStruct[j].address)
		endOfP1 := new(big.Int).Add(address1, ipv6SubnetAddresses(currentResourcesStruct[i].prefix))
		endOfP2 := new(big.Int).Add(address2, ipv6SubnetAddresses(currentResourcesStruct[j].prefix))
		result := endOfP1.Cmp(endOfP2)
		if result >= 0 {
			return false
		}
		return true
	})

	possibleSubnetNum := rootAddressNum
	// iterate over allocated subnets and see if a desired new subnet can be squeezed in
	for _, currentResource := range currentResourcesStruct {

		allocatedSubnetNum, err := Ipv6InetAton(currentResource.address)
		if err != nil {
			return nil, errors.New("Wrong property address: " + currentResource.address + " in current resources")
		}
		chunkCapacity := new(big.Int).Sub(allocatedSubnetNum, possibleSubnetNum)
		if chunkCapacity.Cmp(desiredSize) >= 0 {
			// there is chunk with sufficient capacity between possibleSubnetNum and allocatedSubnet.address
			var newlyAllocatedPrefix = make(map[string]interface{})
			newlyAllocatedPrefix["address"] = Ipv6InetNtoa(possibleSubnetNum)
			newlyAllocatedPrefix["prefix"] = newSubnetMask

			return newlyAllocatedPrefix, nil
		}

		// move possible subnet start to a valid address outside of allocatedSubnet's addresses and continue the search
		possibleSubnetNum, err = findNextFreeSubnetAddress(currentResource, newSubnetMask)
		if err != nil {
			return nil, err
		}
	}

	// check if there is any space left at the end of parent range
	currentAmount := new(big.Int).Add(possibleSubnetNum, newSubnetCapacity)
	rootAmount := new(big.Int).Add(rootAddressNum, rootCapacity)
	if currentAmount.Cmp(rootAmount) < 1 {
		// there sure is some space, use it !
		var newlyAllocatedPrefix = make(map[string]interface{})
		newlyAllocatedPrefix["address"] = Ipv6InetNtoa(possibleSubnetNum)
		newlyAllocatedPrefix["prefix"] = newSubnetMask

		return newlyAllocatedPrefix, nil
	}
	// no suitable range found
	return nil, errors.New("Unable to allocate Ipv6 prefix from: " + rootPrefixStr +
		". Insufficient capacity to allocate a new prefix of size: " + desiredSize.String() +
		". Currently allocated prefixes: " + prefixesToStr(currentResourcesStruct))
}

func calculateDesiredSubnetMask(desiredSize *big.Int) (int, *big.Int) {
	newSubnetBits := big.NewInt(1)
	newSubnet := 1
	for i := 1; i <= 128; i++ {
		newSubnetBits.Mul(newSubnetBits, big.NewInt(2))
		if newSubnetBits.Cmp(desiredSize) >= 0 {
			newSubnet = i
			break
		}
	}
	newSubnetMask := 128 - newSubnet
	newSubnetCapacity := ipv6SubnetAddresses(newSubnetMask)
	return newSubnetMask, newSubnetCapacity
}

// calculate the nearest possible address for a subnet where mask === newSubnetMask
//  that is outside allocatedSubnet
func findNextFreeSubnetAddress(allocatedSubnet Ipv6PrefixStruct, newSubnetMask int) (*big.Int, error) {
	address, err := Ipv6InetAton(allocatedSubnet.address)
	if err != nil {
		return nil, err
	}
	// find the first address after currently iterated allocated subnet
	nextAvailableAddressNum := new(big.Int).Add(address, ipv6SubnetAddresses(allocatedSubnet.prefix))
	// remove any bites from the address above after newSubnetMask
	newSubnetMaskNegative := new(big.Int).Sub(big.NewInt(128), big.NewInt(int64(newSubnetMask)))
	possibleSubnetNum := new(big.Int).Rsh(nextAvailableAddressNum, uint(newSubnetMaskNegative.Int64()))
	possibleSubnetNum.Lsh(possibleSubnetNum, uint(newSubnetMaskNegative.Int64()))
	// keep going until we find an address outside of currently iterated allocated subnet
	for nextAvailableAddressNum.Cmp(possibleSubnetNum) > 0 {
		possibleSubnetNum.Rsh(possibleSubnetNum, uint(newSubnetMaskNegative.Int64()))
		possibleSubnetNum.Add(possibleSubnetNum, big.NewInt(1))
		possibleSubnetNum.Lsh(possibleSubnetNum, uint(newSubnetMaskNegative.Int64()))
	}
	return possibleSubnetNum, nil
}

func getAddressAndPrefixFromCurrentResource(currentResource map[string]interface{}) (string, int, error) {
	value, ok := currentResource["Properties"]
	if !ok {
		return "", 0, errors.New("Unable to extract properties from resource")
	}
	address, ok := value.(map[string]interface{})["address"]
	if !ok {
		return "", 0, errors.New("Unable to extract address resource from properties")
	}
	prefix, ok := value.(map[string]interface{})["prefix"]
	if !ok {
		return "", 0, errors.New("Unable to extract prefix resource from properties")
	}
	prefix, err := NumberToInt(prefix)
	if err != nil {
		return "", 0, err
	}
	return address.(string), prefix.(int), nil
}
